storagenode/piecestore/endpoint_test.go

```diff
@@ -85,3 +91,303 @@ func TestUploadAndPartialDownload(t *testing.T) {
 	t.Log(totalDownload, totalBandwidthUsage.Get, int64(len(planet.StorageNodes))*totalDownload)
 	assert.True(t, totalBandwidthUsage.Get < int64(len(planet.StorageNodes))*totalDownload)
 }
+
+func TestUpload(t *testing.T) {
+	ctx := testcontext.New(t)
+	defer ctx.Cleanup()
+
+	planet, err := testplanet.New(t, 1, 1, 1)
+	require.NoError(t, err)
+	defer ctx.Check(planet.Shutdown)
+
+	planet.Start(ctx)
+
+	client, err := planet.Uplinks[0].DialPiecestore(ctx, planet.StorageNodes[0])
+	require.NoError(t, err)
+
+	for _, tt := range []struct {
+		pieceID storj.PieceID
+		content []byte
+		action  pb.PieceAction
+		err     string
+	}{
+		{ // should successfully store data
+			pieceID: teststorj.PieceIDFromString("99999999999999999999"),
+			content: []byte("xyzwq"),
+			action:  pb.PieceAction_PUT,
+			err:     "",
+		},
+		{ // should err with piece ID not specified
+			pieceID: teststorj.PieceIDFromBytes(make([]byte, 32)),
+			content: []byte("xyzwq"),
+			action:  pb.PieceAction_PUT,
+			err:     "missing piece id",
+		},
+		{ // should successfully store data
+			pieceID: teststorj.PieceIDFromString("99999999999999999999"),
+			content: []byte("xyzwq"),
+			action:  pb.PieceAction_GET,
+			err:     "expected put or put repair action got GET",
+		},
+	} {
+		sum := sha256.Sum256(tt.content)
+		expectedHash := sum[:]
+
+		var serialNumber storj.SerialNumber
+		_, _ = rand.Read(serialNumber[:])
+
+		orderLimit := GenerateOrderLimit(
+			t,
+			planet.Satellites[0].ID(),
+			planet.Uplinks[0].ID(),
+			planet.StorageNodes[0].ID(),
+			tt.pieceID,
+			tt.action,
+			serialNumber,
+			24*time.Hour,
+			24*time.Hour,
+			100,
+		)
+		signer := signing.SignerFromFullIdentity(planet.Satellites[0].Identity)
+		orderLimit, err = signing.SignOrderLimit(signer, orderLimit)
+		require.NoError(t, err)
+
+		uploader, err := client.Upload(ctx, orderLimit)
+		require.NoError(t, err)
+
+		_, err = uploader.Write(tt.content)
+		require.NoError(t, err)
+
+		pieceHash, err := uploader.Commit()
+		if tt.err != "" {
+			require.Error(t, err)
+			require.Contains(t, err.Error(), tt.err)
+		} else {
+			require.NoError(t, err)
+
+			assert.Equal(t, expectedHash, pieceHash.Hash)
+
+			signee := signing.SignerFromFullIdentity(planet.StorageNodes[0].Identity)
+			require.NoError(t, signing.VerifyPieceHashSignature(signee, pieceHash))
+		}
+	}
+}
+
+func TestDownload(t *testing.T) {
+	ctx := testcontext.New(t)
+	defer ctx.Cleanup()
+
+	planet, err := testplanet.New(t, 1, 1, 1)
+	require.NoError(t, err)
+	defer ctx.Check(planet.Shutdown)
+
+	planet.Start(ctx)
+
+	// upload test piece
+	client, err := planet.Uplinks[0].DialPiecestore(ctx, planet.StorageNodes[0])
+	require.NoError(t, err)
+
+	var serialNumber storj.SerialNumber
+	_, _ = rand.Read(serialNumber[:])
+
+	orderLimit := GenerateOrderLimit(
+		t,
+		planet.Satellites[0].ID(),
+		planet.Uplinks[0].ID(),
+		planet.StorageNodes[0].ID(),
+		teststorj.PieceIDFromString("existing-piece-id"),
```

Comment:
You can use `storj.PieceID{1,2,3,4,5}`... the `teststorj.PieceIDFromString`, isn't that great.

Reply:
done